### PR TITLE
Building an Artificial Neural Network with Keras

### LINKED ISSUE #2390: Building an Artificial Neural Network with TensorFlow 2
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of the article
Your title should be descriptive of the article/tutorial. Be Specific. *Use keyword research to gain your article a higher ranking.*

Building an Artificial Neural Network with TensorFlow 2

### Introduction paragraph (2-3 paragraphs):
Please write the Introductory paragraph(s), that would be included in your article. We will use this writing snippet to help us assess overall quality before approval. We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.
Take your time and write the content as you would intend to get it published. 

In this guide, we will learn how to build an Artificial Neural Network (ANN) with TensorFlow 2.
One application of machine learning in business is predicting customers that will churn. If we know the customers that will churn, we can provide special services to them. We do not want to lose any customers.
We will build and train an Artificial Neural Network to predict customers that will churn from a bank using the customer-churning dataset from Kaggle. Some features of the dataset we will use are credit score, country, gender, age, tenure, balance, etc. 
After reading this guide, you will be able to build an ANN to solve binary classification problems with any dataset.

### Key takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article? Use this area to get your ideas down on the bulk of your article or tutorial.

- Label Encoding & One-Hot Encoding Categorical Variables,
- Build and visualize model architecture with Keras visualizer,
- Compile and train an ANN with Keras on top of TensorFlow 2,
- Make Predictions on a new customer,
- Check the model's accuracy score and create a confusion matrix.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

- https://keras.io/api/

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Inyrkz**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**

This fixes #2390.
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Inyrkz** [comment]
This fixes #2390

**Comment by Bayler** [comment]
Hi @Inyrkz I will be reviewing your article.

Please make sure you have checked all the boxes in the article checklist before submission. In that case, please upload a hero image.

**Comment by Bayler** [comment]
@Inyrkz Thank you for uploading your hero image. Please note that hero images are named with hero.jpg or hero.png etc. The same way your main file should always be index.md

(hero-image) isn't correct. Please rename it.

**Comment by Bayler** [comment]
@Inyrkz Please provide a link to your code implementation on [Replit](www.replit.com) so that I can confirm your code and output.

**Comment by Inyrkz** [comment]
Can I share a copy of the Colab notebook instead?

**Comment by Bayler** [comment]
> Can I share a link to a copy of the Colab notebook instead?

Sure @Inyrkz 

You can also include the link to your notebook at the bottom of your article.

For example,

 "You can check the full code here"

**Comment by Inyrkz** [comment]
https://colab.research.google.com/drive/1dpgT_qZL2WsLSxzHF3142IEQN7-biOrw?authuser=1#scrollTo=ZaTwK7ojXr2F

I added the link to the notebook in my github repo at top of the article.

**Comment by Bayler** [comment]
> https://colab.research.google.com/drive/1dpgT_qZL2WsLSxzHF3142IEQN7-biOrw?authuser=1#scrollTo=ZaTwK7ojXr2F
> 
> I added the link to the notebook in my github repo at top of the article.

Noted ✅

I'll check it out

**Comment by hectorkambow** [comment]
👋 Good afternoon - Hector here.
I'll be running a final review on your PR when ready.
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Building an Artificial Neural Network with Keras #2762.pdf](https://github.com/section-engineering-education/engineering-education/files/6771696/Building.an.Artificial.Neural.Network.with.Keras.2762.pdf)

**Comment by Inyrkz** [comment]
Hi!
I have updated the article. I hope it is better now.

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/build-ann-with-keras/index.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+---  
+
+![Hero Image](/engineering-education/build-ann-with-keras/hero.PNG)
+
+In this article, you will learn how to build and train an artificial neural network with Keras. We will make a model that will tell us if a customer will churn. That can be very useful in businesses. If you know the customers that will churn, you can provide these customers with better offers. So you can keep them. We will use machine learning to determine customers that are likely to churn. We have a sample dataset from a bank. We will predict the customers that will stop banking with this bank. Here is the GitHub [repo](https://github.com/Inyrkz/Customer-Churn) for this project.
+
+### Prerequisites
+-	Basics of [Artificial Neural Network](https://towardsdatascience.com/neural-networks-basics-29cc093b82be)
+-	[Google Colab](https://colab.research.google.com/) or Jupyter Notebook
+-	Churn modeling dataset from [Kaggle](https://www.kaggle.com/adammaus/predicting-churn-for-bank-customers)
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:25]: 1. Let us try and promote our own articles on the section engineering website. In this regard, we have plenty of articles on ANN the Section articles section. Reference that instead. Let's promote readers to read our own articles.
2. Indicate whether you've used Google Colab or Jupyter Notebook.
For example, "In this tutorial, we have used Google Colab". Let it be clear.

File: content/articles/build-ann-with-keras/index.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+### Outline
+
+-	Import Libraries
+-	Data Pre-processing
+-	Build and Visualize Model
+-	Train Model
+-	Make Predictions on New Customer
+
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:34]: Please add links to the respective headers inside the article. Use this PR as reference (https://github.com/section-engineering-education/engineering-education/pull/2785/files)

File: content/articles/build-ann-with-keras/index.md (hunk 1 above)
Comment (Inyrkz) [reply content/articles/build-ann-with-keras/index.md]: Done

File: content/articles/build-ann-with-keras/index.md (hunk 2 above)
Comment (Inyrkz) [reply content/articles/build-ann-with-keras/index.md]: Links added

File: content/articles/build-ann-with-keras/index.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+print(y)
+```
+Here are the features and labels obtained after separation:
+
+![features of dataset](/engineering-education/build-ann-with-keras/features.PNG)
+
+![dataset label](/engineering-education/build-ann-with-keras/labels.PNG)
+
+You notice that there are some categorical variables in our dataset. They are in the geography and gender columns. We have to encode these variables. Since there are two unique variables in the gender column, we label-encode it. Then, we one-hot encode the geography column.
+
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:95]: Explain briefly the differences between label-encoding and label encoding. This is beneficial to your readers.

File: content/articles/build-ann-with-keras/index.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+Most of the libraries we will be using have been pre-installed on Google Colab. So, we import them into our code.
+
+```python
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+```
+**Explain what the code below does and why it is necessary**
+
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:46]: I've left a comment here in bold. Replace it with the necessary instruction.

File: content/articles/build-ann-with-keras/index.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+```python
+y_pred = ann.predict(X_test)
+y_pred = (y_pred > 0.5)
+pd.DataFrame(list(zip(y_test, y_pred)), columns=['Actual', 'Predicted'])
+```
+![Actual Values vs Predicted Values](/engineering-education/build-ann-with-keras/actual-vs-predicted-values.PNG)
+
+**Comment on these results obtained**
+
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:229]: I've left a comment here in bold. Replace it with the necessary instruction.

File: content/articles/build-ann-with-keras/index.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+```python
+from sklearn.metrics import confusion_matrix, accuracy_score
+print(confusion_matrix(y_test, y_pred))
+print(accuracy_score(y_test, y_pred))
+```
+![Confusion Matrix & Accuracy Score](/engineering-education/build-ann-with-keras/confusion-matrix-and-accuracy-score.PNG)
+
+**Comment on these scores obtained**
+
```
Comment (Bayler) [inline content/articles/build-ann-with-keras/index.md:240]: I've left a comment here in bold. Replace it with the necessary instruction.

**Comment by Bayler** [comment]
@Inyrkz Hi. This was a nice read. Just a few places that I feel you need to add more information then we are good to go for final review.

**Comment by hectorkambow** [comment]
@Inyrkz 
Just wanted to get a status update on this article - as were trying to clear the queue wherever possible. 👍

File: content/articles/build-ann-with-keras/index.md (hunk 5 above)
Comment (Inyrkz) [reply content/articles/build-ann-with-keras/index.md]: Done!

**Review by ahmadmardeni1 — COMMENTED** [review]
Great article @Inyrkz 🚀

Thank you for the review @Bayler 

Please go over the changes I made. When the output is simple, like a number or text, then there is no need to put a screenshot.
